src/solver/conv_hip_implicit_gemm_bwd_data_v4r1_xdlops.cpp

```diff
@@ -0,0 +1,690 @@
+                                                 GetEPackLength(ctx, true));
+    return lds_size <= 64 * 1024;
+}
+
+PerformanceImplicitGemmBwdDataV4R1Xdlops::PerformanceImplicitGemmBwdDataV4R1Xdlops(bool spare)
+{
+    GemmNPerBlock = spare ? 16 : 64;
+    GemmMPerBlock = spare ? 4 : 64;
+    GemmKPerBlock = spare ? 4 : 8;
+
+    GemmMPerWave = spare ? 4 : 64;
+    GemmNPerWave = spare ? 16 : 64;
+
+    InBlockCopyClusterLengths_GemmK = 4;
+    InBlockCopyClusterLengths_GemmN = 4;
+
+    WeiBlockCopyClusterLengths_GemmK = 2;
+    WeiBlockCopyClusterLengths_GemmM = 4;
+
+    use_spare_set = spare;
+}
+
+PerformanceImplicitGemmBwdDataV4R1Xdlops::PerformanceImplicitGemmBwdDataV4R1Xdlops(
+    int BPerBlock_,
+    int KPerBlock_,
+    int EPerBlock_,
+    int GemmMPerWave_,
+    int GemmNPerWave_,
+    int InBlockCopyClusterLengths_E_,
+    int InBlockCopyClusterLengths_B_,
+    int WeiBlockCopyClusterLengths_E_,
+    int WeiBlockCopyClusterLengths_K_,
+    bool use_spare_set_)
+    : GemmNPerBlock(BPerBlock_),
+      GemmMPerBlock(KPerBlock_),
+      GemmKPerBlock(EPerBlock_),
+      GemmMPerWave(GemmMPerWave_),
+      GemmNPerWave(GemmNPerWave_),
+      InBlockCopyClusterLengths_GemmK(InBlockCopyClusterLengths_E_),
+      InBlockCopyClusterLengths_GemmN(InBlockCopyClusterLengths_B_),
+      WeiBlockCopyClusterLengths_GemmK(WeiBlockCopyClusterLengths_E_),
+      WeiBlockCopyClusterLengths_GemmM(WeiBlockCopyClusterLengths_K_),
+      use_spare_set(use_spare_set_)
+{
+}
+
+bool PerformanceImplicitGemmBwdDataV4R1Xdlops::
+operator==(const PerformanceImplicitGemmBwdDataV4R1Xdlops& other) const
+{
+    // clang-format off
+    return GemmNPerBlock == other.GemmNPerBlock
+        && GemmMPerBlock == other.GemmMPerBlock
+        && GemmKPerBlock == other.GemmKPerBlock
+        && GemmMPerWave == other.GemmMPerWave
+        && GemmNPerWave == other.GemmNPerWave
+        && InBlockCopyClusterLengths_GemmK == other.InBlockCopyClusterLengths_GemmK
+        && InBlockCopyClusterLengths_GemmN == other.InBlockCopyClusterLengths_GemmN
+        && WeiBlockCopyClusterLengths_GemmK == other.WeiBlockCopyClusterLengths_GemmK
+        && WeiBlockCopyClusterLengths_GemmM == other.WeiBlockCopyClusterLengths_GemmM
+        && use_spare_set == other.use_spare_set;
+    // clang-format on
+}
+
+bool PerformanceImplicitGemmBwdDataV4R1Xdlops::IsValidValue() const
+{
+    // clang-format off
+    return IsTwoPower<16,128>(GemmNPerBlock)
+        && IsTwoPower<4,128>(GemmMPerBlock)
+        && IsTwoPower<4,32>(GemmKPerBlock)
+        && IsTwoPower<4,64>(GemmMPerWave)
+        && IsTwoPower<16,64>(GemmNPerWave)
+        && IsTwoPower<4,16>(InBlockCopyClusterLengths_GemmK)
+        && IsTwoPower<4,64>(InBlockCopyClusterLengths_GemmN)
+        && IsTwoPower<2,16>(WeiBlockCopyClusterLengths_GemmK)
+        && IsTwoPower<4,128>(WeiBlockCopyClusterLengths_GemmM); // clang-format on
+}
+
+bool PerformanceImplicitGemmBwdDataV4R1Xdlops::SetNextValue()
+{
```

Comment:
Could you please reduce search space by eliminating ClusterLengths variables?

Take a look at solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_fwd_fp32.cpp. @zjing14 made those changes in fwd fp32. You could pursue similar strategy and reduce tuning time. I will eliminate 4 tuning variables.

Reply:
Hi, @TejashShah @zjing14, which number could be picked for the ClusterLengths variables? Do you have some heuristic data for these 4 params?

Reply:
I removed 4 tuning variables.